pkg/issuer/acme/setup.go

```diff
@@ -37,35 +37,55 @@ import (
 	cmmeta "github.com/jetstack/cert-manager/pkg/apis/meta/v1"
 	logf "github.com/jetstack/cert-manager/pkg/logs"
 	"github.com/jetstack/cert-manager/pkg/util/errors"
-	"github.com/jetstack/cert-manager/pkg/util/kube"
 	"github.com/jetstack/cert-manager/pkg/util/pki"
 )
 
 const (
 	errorAccountRegistrationFailed = "ErrRegisterACMEAccount"
 	errorAccountVerificationFailed = "ErrVerifyACMEAccount"
 	errorAccountUpdateFailed       = "ErrUpdateACMEAccount"
+	errorInvalidConfig             = "InvalidConfig"
+	errorInvalidURL                = "InvalidURL"
 
 	successAccountRegistered = "ACMEAccountRegistered"
 	successAccountVerified   = "ACMEAccountVerified"
 
-	messageAccountRegistrationFailed = "Failed to register ACME account: "
-	messageAccountVerificationFailed = "Failed to verify ACME account: "
-	messageAccountUpdateFailed       = "Failed to update ACME account:"
-	messageAccountRegistered         = "The ACME account was registered with the ACME server"
-	messageAccountVerified           = "The ACME account was verified with the ACME server"
+	messageAccountRegistrationFailed     = "Failed to register ACME account: "
+	messageAccountVerificationFailed     = "Failed to verify ACME account: "
+	messageAccountUpdateFailed           = "Failed to update ACME account:"
+	messageAccountRegistered             = "The ACME account was registered with the ACME server"
+	messageAccountVerified               = "The ACME account was verified with the ACME server"
+	messageNoSecretKeyGenerationDisabled = "the ACME issuer config has 'disableAccountKeyGeneration' set to true, but the secret was not found: "
+	messageInvalidPrivateKey             = "Account private key is invalid: "
+
+	messageTemplateUpdateToV2              = "Your ACME server URL is set to a v1 endpoint (%s). You should update the spec.acme.server field to %q"
+	messageTemplateNotRSA                  = "ACME private key in %q is not of type RSA"
+	messageTemplateFailedToParseURL        = "Failed to parse existing ACME server URI %q: %v"
+	messageTemplateFailedToParseAccountURL = "Failed to parse existing ACME account URI %q: %v"
 )
 
 // Setup will verify an existing ACME registration, or create one if not
 // already registered.
 func (a *Acme) Setup(ctx context.Context) error {
 	log := logf.FromContext(ctx)
 
+	// Correct reason and message for issuer's Ready condition must be always set
+	// before returning from this function. Status must be set if not false.
+	status := cmmeta.ConditionFalse
+	var reason, msg string
+	defer func() {
+		apiutil.SetIssuerCondition(a.issuer,
+			a.issuer.GetGeneration(),
+			v1.IssuerConditionReady,
+			status,
+			reason,
+			msg)
+	}()
```

Comment:
Agree that this could be improved.
If the condition was set on the issuer in `Sync` though I guess we'd still need to build the condition in `Setup` and return it?

I do find it a little confusing that we set the condition on the Issuer in `Setup`, but actually apply that update in `Sync` [here](https://github.com/jetstack/cert-manager/blob/master/pkg/controller/issuers/sync.go#L46).